### PR TITLE
ncm-metaconfig:  Support all combinations of udev rule creation

### PR DESCRIPTION
**Pull request by dj0101**
Currently supported udev template rules are limited and barely usable for
    real udev usage. New schema and template can support all the combinations
    of the match and assign keys creating any rules

Description of PR and copyright notice:

Current udev templating is not sufficient enough to handle various different usage and the new commit will allow generation of all the combination of udev rules generations currently supported.  Since this is templating addition, anyone wishes to use the previous template are free to do so, the new addition of the templates just handles all situations whereas existing templates can handle very specific cases only. 

**A copyright notice**
THIS SOFTWARE IS CONTRIBUTED SUBJECT TO THE TERMS OF THE APACHE LICENSE, V.2.0. YOU MAY OBTAIN A COPY OF THE LICENSE AT https://www.apache.org/licenses/LICENSE-2.0.
THIS SOFTWARE IS LICENSED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE AND ANY WARRANTY OF NON-INFRINGEMENT, ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. THIS SOFTWARE MAY BE REDISTRIBUTED TO OTHERS ONLY BY EFFECTIVELY USING THIS OR ANOTHER EQUIVALENT DISCLAIMER IN ADDITION TO ANY OTHER REQUIRED LICENSE TERMS.

### CODE REVIEW

**Comment by jrha** [comment]
@hpcugentbot test this please

**Comment by dj0101** [comment]
Hi, could i get someone to review the PR please? 

Thanks

**Comment by dj0101** [comment]
BUMP for review

**Comment by dj0101** [comment]
Bump

**Comment by jrha** [comment]
Sorry, I'll try to get this reviewed next week.

**Comment by dj0101** [comment]
bump!

**Comment by dj0101** [comment]
> Sorry, I'll try to get this reviewed next week.

Hi any luck on the review? 

Thanks

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule.pan [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+       - dict of list of dict allows mulitple matching or multiple assign with 
+         different keys (example of this below, 'attrs[0]' attribute or 'run[0]'
+         assign attribute)
+   - match_rule attributes could have operator of 2 types, ==, and != 
+   - assign_rule attributes could have operator of 3 types, =, := and += 
+
+Refer to manpage udev(7) for keys.
+            
+Rules to generate:
+#1 rule: ACTION=="add|change", ATTRS{device}=="0x1572", ATTRS{vendor}=="0x8086", DRIVERS=="i40e", \
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule.pan:18]: Ah slight snag which I didn't think of, `}` ends the annotation, so everything after this is a syntax error.

Comment (jrha) [reply ncm-metaconfig/src/main/metaconfig/udev/pan/rule.pan:18]: `&lbrace;vendor&rbrace;` would be accepted and make it into the documentation at the expense of code readability

Comment (dj0101) [reply ncm-metaconfig/src/main/metaconfig/udev/pan/rule.pan]: updated as suggested

Comment (dj0101) [reply ncm-metaconfig/src/main/metaconfig/udev/pan/rule.pan]: not familiar with the tests being failed, could you help ?

Comment (dj0101) [reply ncm-metaconfig/src/main/metaconfig/udev/pan/rule.pan]: @jrha  any chance you can help with the testing?

**Comment by ned21** [comment]
@dj0101  - the failed test is this one - GH highlights the wrong part of the output:

```
# Looks like you failed 1 test of 96.
src/test/perl/service-udev.t ............. 
ok 1 - service udev defined for ttpath
...
not ok 11 - No invalid pan templates udev/pan/rule.pan
ok 12 - Found schema udev/pan/schema.pan
```

You then have to scroll up to find the relevant output, around line 48462:

```
# VERBOSE found 5 pan templates: udev/pan/action_attr.pan, udev/pan/blockdev_run.pan, udev/pan/rule_schema/schema.pan, udev/pan/schema.pan, udev/pan/network_config.pan
#   Failed test 'No invalid pan templates udev/pan/rule.pan'
#   at /__w/configuration-modules-core/configuration-modules-core/ncm-metaconfig/target/dependency/build-scripts/Test/Quattor/TextRender.pm line 391.
#          got: '1'
#     expected: '0'
# VERBOSE Pan source /__w/configuration-modules-core/configuration-modules-core/ncm-metaconfig/src/main/metaconfig/udev/pan/action_attr.pan from relative udev/pan/action_attr.pan dest 
```

This is some weird meta-test but it looks like it wants a file called udev/pan/rule which you've added. Are you sure this is valid PAN? I see two potential issues:
1. There's a space before template on the first line
2. It has no PAN code in it, only an unterminated annotation.

So I think you need to add a } to terminate the annotation. And if that doesn't work then revert to using comments since if that works since clearly annotations are built for this level of complexity!

**Comment by dj0101** [comment]
> 

the problem seemed to be missing unique statement before the template.  It looks like the testing has completed without any problem now.

**Comment by dj0101** [comment]
> > 
> 
> the problem seemed to be missing unique statement before the template. It looks like the testing has completed without any problem now.

@jrha Hi, looks like all checks have passed. Could you do your review ?

**Comment by dj0101** [comment]
bump!

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+equality check.
+matched key result in action/assigment in ether dict of dict with a key or without one that has a
+ value and an operator with assignment.
+
+There could be multiple lines of rules with different matches independent of each rules.
+
+example:
+/etc/udev/rules.d/00-x710.rules
+
+ACTION=="add|change", ATTRS{device}=="0x1572", RUN+="/sbin/ethtool -s '%k' speed 10000"
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:19]: Hmm, the `}` here terminates the annotation, everything after it is treated as pan code, which then fails with a parse error and there doesn't appear to be any way to escape this.

Perhaps as @ned21 suggested this is actually too complex for an annotation.

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+type udev_rule_assign_operator = string with match(SELF, '^(=|\+=|:=)$');
+
+type string_match_structure = {
+    @{ dict of dict with string key, eg ACTION=="add|change" }
+    'value' : string
+    'operator' : udev_rule_match_operator
+};
+
+type dict_match_structure = {
+    @{ dict of dict with a key, eg ATTRS{device}=="0x1572" }
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:40]: Same here, maybe document the usage instead?
```suggestion
    @{
        dict of dict with a key, eg
        dict(
            'key', 'device',
            'operator', '==', 
            'value', '0x1572',
        );
    }
```

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+};
+
+type string_assign_structure = {
+    @{ dict of dict without key, eg RUN+="/sbin/ethtool --set-priv-flags '%k'" }
+    'value' : string
+    'operator' : udev_rule_assign_operator
+};
+
+type dict_assign_structure = {
+    @{ dict of dict with a key, eg ATTR{vendor}="0x8086" }
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:53]: and here:
```suggestion
    @{ dict of dict with a key } # eg ATTR{vendor}="0x8086"
```

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+ACTION is one of the match keys with value of add|change with '==' operator.
+RUN is a matched key result of all the matched keys, ACTION, and ATTRS being matched, with value
+of "/sbin/ethtool ..." and operator of '+='.
+
+Please refer to udev(7) man page for the list of match and assignment or action
+}
+
+type udev_rule_match_operator = string with match(SELF, '^(={2}|!=)$');
+
+type udev_rule_assign_operator = string with match(SELF, '^(=|\+=|:=)$');
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:31]:
```suggestion
type udev_rule_assign_operator = choice('=', '+=', ':=');
```

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+
+ACTION is one of the match keys with value of add|change with '==' operator.
+RUN is a matched key result of all the matched keys, ACTION, and ATTRS being matched, with value
+of "/sbin/ethtool ..." and operator of '+='.
+
+Please refer to udev(7) man page for the list of match and assignment or action
+}
+
+type udev_rule_match_operator = string with match(SELF, '^(={2}|!=)$');
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:29]:
```suggestion
type udev_rule_match_operator = choice('==', '!=');
```

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+type string_match_structure = {
+    @{ dict of dict with string key, eg ACTION=="add|change" }
+    'value' : string
+    'operator' : udev_rule_match_operator
+};
+
+type dict_match_structure = {
+    @{ dict of dict with a key, eg ATTRS{device}=="0x1572" }
+    'key' : string
+    'value' : string
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:42]:
```suggestion
    'key' : string_trimmed
    'value' : string_trimmed
```
Are there any instances where `value` wouldn't be a hexadecimal value?

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    'operator' : udev_rule_match_operator
+};
+
+type string_assign_structure = {
+    @{ dict of dict without key, eg RUN+="/sbin/ethtool --set-priv-flags '%k'" }
+    'value' : string
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:48]:
```suggestion
    'value' : string_trimmed
```

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+    'operator' : udev_rule_assign_operator
+};
+
+type dict_assign_structure = {
+    @{ dict of dict with a key, eg ATTR{vendor}="0x8086" }
+    'key' : string
+    'value' : string
```
Comment (jrha) [inline ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:55]:
```suggestion
    'key' : string_trimmed
    'value' : string_trimmed
```

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan (hunk 4 above)
Comment (jrha) [reply ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:53]:
```suggestion
    @{
        dict of dict with a key, eg
        dict(
            'key', 'vendor',
            'operator', '=', 
            'value', '0x8086',
        );
    }
```

File: ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan (hunk 2 above)
Comment (jrha) [reply ncm-metaconfig/src/main/metaconfig/udev/pan/rule_schema/schema.pan:19]:
```suggestion
ACTION=="add|change", ATTRS{device}=="0x1572", RUN+="/sbin/ethtool -s '%k' speed 10000"

'rule' = dict(
    'match_rule', dict(
        'attrs', list(
            dict(
                'key', 'device',
                'operator', '==', 
                'value', '0x1572',
            ),
        ),
    ),
    'udev_assign_rule': dict(
        'run', list(
             dict(
                'operator', '+=', 
                'value', "/sbin/ethtool -s '%k' speed 10000",
            ),
        ),
    ),
);
```

**Review by jrha — APPROVED** [review]

**Pull request merged**